src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -62,24 +98,28 @@ public bool StartAssignment(HostAssignmentContext context)
                 return false;
             }
 
-            if (context.IsWarmupRequest)
-            {
-                // Based on profiling download code jit-ing holds up cold start.
-                // Pre-jit to avoid paying the cost later.
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
-                return true;
-            }
-            else if (_assignmentContext == null)
+            return true;
+        }
+
+        private async Task AssignAsync(HostAssignmentContext assignmentContext)
+        {
+            if (assignmentContext.IsWarmupRequest)
             {
-                lock (_assignmentLock)
+                try
+                {
+                    await DownloadWarmupAsync(assignmentContext.GetRunFromPkgContext());
+                }
+                catch (Exception ex)
                 {
-                    if (_assignmentContext != null)
-                    {
-                        return _assignmentContext.Equals(context);
-                    }
-                    _assignmentContext = context;
+                    _logger.LogError(ex, "Warmup download failed");
+                    await _meshServiceClient.NotifyHealthEvent(ContainerHealthEventType.Warning, GetType(), "Warmup download failed");
+                    throw;
```

Comment:
There wasn't any error handling for this previously - mostly tried to match the same pattern as an assignment failure, but not sure if this should be as aggressive (`ContainerHealthEventType.Fatal` and throw).